zerver/lib/streams.py

```diff
@@ -1005,3 +1008,42 @@ def get_subscribed_private_streams_for_user(user_profile: UserProfile) -> QueryS
         .filter(subscribed=True)
     )
     return subscribed_private_streams
+
+
+@transaction.atomic
+def update_stream_active_status_for_realm(realm: Realm, date_days_ago: datetime) -> int:
+    active_stream_ids = (
+        Message.objects.filter(
+            date_sent__gte=date_days_ago, recipient__type=Recipient.STREAM, realm=realm
+        )
+        .values_list("recipient__type_id", flat=True)
+        .distinct()
+    )
+    streams_to_mark_inactive = Stream.objects.filter(is_recently_active=True, realm=realm).exclude(
+        id__in=active_stream_ids
+    )
```

Comment:
Why query plan does this generate?

Reply:
Ran `streams_to_mark_inactive.explain()`:

```bash
Seq Scan on zerver_stream  (cost=13.69..15.01 rows=1 width=162)
  Filter: (is_recently_active AND (NOT (hashed SubPlan 1)) AND (realm_id = 1))
  SubPlan 1
    ->  Unique  (cost=13.68..13.69 rows=1 width=4)
          ->  Sort  (cost=13.68..13.69 rows=1 width=4)
                Sort Key: u2.type_id
                ->  Nested Loop  (cost=0.30..13.67 rows=1 width=4)
                      ->  Index Scan using zerver_message_realm_id_849a39c8 on zerver_message u0  (cost=0.14..4.16 rows=1 width=4)
                            Index Cond: (realm_id = 1)
                            Filter: (date_sent >= '2024-04-10 13:23:21.408525+00'::timestamp with time zone)
                      ->  Index Scan using zerver_recipient_pkey on zerver_recipient u2  (cost=0.15..8.17 rows=1 width=8)
                            Index Cond: (id = u0.recipient_id)
                            Filter: (type = 2)
Seq Scan on zerver_stream  (cost=56.72..58.04 rows=4 width=162)
  Filter: (is_recently_active AND (NOT (hashed SubPlan 1)) AND (realm_id = 2))
  SubPlan 1
    ->  Unique  (cost=0.42..56.72 rows=1 width=4)
          ->  Nested Loop  (cost=0.42..56.72 rows=1 width=4)
                ->  Index Scan using zerver_recipient_type_type_id_29712363_uniq on zerver_recipient u2  (cost=0.15..36.35 rows=11 width=8)
                      Index Cond: (type = 2)
                ->  Index Only Scan using zerver_message_realm_recipient_date_sent on zerver_message u0  (cost=0.27..1.81 rows=4 width=4)
                      Index Cond: ((realm_id = 2) AND (recipient_id = u2.id) AND (date_sent >= '2024-04-10 13:23:21.408525+00'::timestamp with time zone))
Seq Scan on zerver_stream  (cost=13.69..15.01 rows=1 width=162)
  Filter: (is_recently_active AND (NOT (hashed SubPlan 1)) AND (realm_id = 3))
  SubPlan 1
    ->  Unique  (cost=13.68..13.69 rows=1 width=4)
          ->  Sort  (cost=13.68..13.69 rows=1 width=4)
                Sort Key: u2.type_id
                ->  Nested Loop  (cost=0.30..13.67 rows=1 width=4)
                      ->  Index Scan using zerver_message_realm_id_849a39c8 on zerver_message u0  (cost=0.14..4.16 rows=1 width=4)
                            Index Cond: (realm_id = 3)
                            Filter: (date_sent >= '2024-04-10 13:23:21.408525+00'::timestamp with time zone)
                      ->  Index Scan using zerver_recipient_pkey on zerver_recipient u2  (cost=0.15..8.17 rows=1 width=8)
                            Index Cond: (id = u0.recipient_id)
                            Filter: (type = 2)
2024-10-07 13:23:21.495 INFO [zerver.management.commands.update_stream_active_status] Marked 5 streams as not recently active.
```

Reply:
OK those sequential scans are bad. I think we need a new index on `(realm, is_recently_active)` -- that should make those two sequential scans able to be index scans on that index. Or maybe we need `deactivated` in the list? Let me see.

Reply:
Hmm. If not for the issue of moving messages, we could potentially do great with the `StreamCount` analytics table. Basically only consider for marking inactive those channels that do not have records in that table.

```
        CountStat(                                                                                      
            "messages_in_stream:is_bot:day",                                                            
            sql_data_collector(                                                                         
                StreamCount, count_message_by_stream_query(realm), (UserProfile, "is_bot")              
            ),                                                                                          
            CountStat.DAY,                                                                              
        ),                                                                                              
```

I suppose we could have an algorithm for checking a given realm like this:

- First, start with selecting (0) non-deactivated channels in the organization that are (1) marked as `recently_active` and (2) nonzero entries in `StreamCount` within the relevant date window. This would let us avoid doing work for channels where any messages have been sent recently.
- Then, for each such channel, just do a direct query for whether the channel has any messages (which would have to have gotten there via moving topics). These individual `.exists()` queries would be very fast thanks to `zerver_message_realm_recipient_date_sent`. And for channels where there are indeed no recent messages, the query should complete immediately (<1ms of work), and we set `recently_active=False` and then don't need to process this channel again.

Then the only channels for which we do work every cron job run are those whose recent activity is exclusively moved messages, which should be quite rare, given that people moving messages generally will generate a notification bot message about the move, or a human reply.

Notably this would mean that the cron job will skip processing channels where moves/deletion have removed all messages in the channel; those will stay "active" incorrectly. If we think that's the wrong product behavior, the fix would be to do a quick `.exists()` check in those code paths. I think we'd be able to export the function for doing the `exists` check as an abstraction to call from such code paths, next to `check_update_first_message_id`?